### PR TITLE
feat(core): accept a record in all() / allAsync()

### PR DESCRIPTION
**Pull request by btravers**
Adds a **record** overload to `all` and `allAsync` (idea from byethrow's `collect`/`sequence`, adapted to unthrown's first-`Err`-wins semantics).

```ts
all({ id: ok(1), name: ok("ada") });
// Result<{ id: number; name: string }, never>

await allAsync({ user: loadUser(), org: loadOrg() });
// AsyncResult<{ user: User; org: Org }, NotFound>
```

**Output mirrors input shape** — tuple → tuple (positional), array → `T[]`, record → record of values. The folding rules are unchanged: first `Err` short-circuits, any `Defect` dominates. This is **not** error accumulation (that stays deliberately excluded) — just an ergonomic shape for named parallel work, no tupling.

- New shared `foldAll` helper backs both array and record forms.
- Tests for record success / first-Err / Defect-dominates / empty record, sync and async.
- Docs (`core-concepts`) + `CLAUDE.md` aggregate spec updated; 100% core line/function coverage held.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: docs/guide/core-concepts.md [hunk 1]
```diff
@@ -102,17 +102,23 @@ ok(1).match({ ok, err, defect }); // fold all three channels
...
+mirrors the input shape — a fixed tuple keeps its positional types, a dynamic
+`Result<T, E>[]` collapses to `Result<T[], E>`, and a **record** becomes a record
+of values (handy for named parallel work, no tupling):
 
 ```ts
 import { all, ok, type Result } from "unthrown";
 
 all([ok(1), ok("two"), ok(true)]).unwrap(); // [1, "two", true] (typed [number, string, boolean])
 all([ok(1), ok(2)] as Result<number, never>[]).unwrap(); // number[]
+all({ id: ok(1), name: ok("ada") }).unwrap(); // { id: 1, name: "ada" }
```
Comment (btravers) [inline docs/guide/core-concepts.md]: I think it would be better to have separate function `all` and `allFromDict`

Comment (btravers) [reply docs/guide/core-concepts.md]: Done — split into separate functions. `all` / `allAsync` stay tuple/array-only; added `allFromDict` / `allFromDictAsync` for the record form. Agreed that positional vs named is a distinct concept, so this fits the "one name per concept" rule better than an overload.

Also from this review pass: the record fold now writes keys via `Object.defineProperty` (no `__proto__` pollution), and I added the missing async-record parity tests (Defect-dominates, empty record) + a pollution test.

File: packages/core/src/interop.ts [hunk 2]
```diff
@@ -179,75 +179,123 @@ type AllOk<
...
+    return firstDefect ?? firstErr ?? ok(values);
+  }
+
+  const values: Record<string, unknown> = {};
+  for (const [key, r] of Object.entries(results)) {
+    if (r.tag === "Defect") firstDefect ??= r;
+    else if (r.tag === "Err") firstErr ??= r;
+    else values[key] = r.value;
+  }
+  return firstDefect ?? firstErr ?? ok(values);
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/core/src/interop.ts:227]: Fixed in ccfe1dd — the record fold now writes keys with `Object.defineProperty` (and the async path pairs keys on a null-proto accumulator), so a `__proto__` key lands as a normal own property instead of mutating the prototype. Added a pollution regression test.

File: packages/core/src/interop.ts [hunk 3]
```diff
@@ -179,75 +179,123 @@ type AllOk<
...
+    return new AsyncRes(settled);
+  }
+  const entries = Object.entries(results);
+  const settled = Promise.all(entries.map(([, ar]) => ar)).then((resolved) => {
+    const byKey: ResultRecord = {};
+    entries.forEach(([key], i) => {
+      byKey[key] = resolved[i] as Result<unknown, unknown>;
+    });
+    return foldAll(byKey);
+  });
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/core/src/interop.ts:335]: Fixed in ccfe1dd — same defineProperty treatment; the async `byKey` accumulator is now `Object.create(null)`, so pairing resolved values back to keys can't pollute either.

File: packages/core/src/interop.ts [hunk 4]
```diff
@@ -179,75 +179,123 @@ type AllOk<
...
- *
- * @typeParam Rs - the tuple/array of input `AsyncResult` types.
- * @param results - the async results to combine.
+ * The inputs are resolved **concurrently** (order preserved); the resolved
+ * `Result`s are then folded with the same rules as {@link all} — first `Err`
+ * short-circuits, any `Defect` dominates. As ever, the returned `AsyncResult`'s
+ * internal promise never rejects. The output mirrors the input shape: a fixed
+ * tuple keeps positional types, a dynamic array `AsyncResult<T, E>[]` collapses
+ * to `AsyncResult<T[], E>`, and a record `{ a, b }` becomes
+ * `AsyncResult<{ a; b }, E>`.
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/core/src/interop.ts]: Resolved by the refactor in ccfe1dd — record aggregation moved to dedicated `allFromDict`/`allFromDictAsync` functions whose JSDoc shows a proper `{ a: A; b: B }` shape; the unclear inline example is gone.

File: CLAUDE.md [hunk 5]
```diff
@@ -94,10 +94,14 @@ async work re-enters via `fromPromise` / `fromSafePromise` and composes with
...
-  are **tuple-or-array**: a fixed tuple keeps positional types, a dynamic
-  `Result<T, E>[]` / `AsyncResult<T, E>[]` collapses to `Result<T[], E>` /
-  `AsyncResult<T[], E>` with no cast. `allAsync` resolves its inputs concurrently
-  (order preserved) and never rejects.
+  are **tuple-, array-, or record-shaped**, and the output mirrors the input: a
+  fixed tuple keeps positional types, a dynamic `Result<T, E>[]` /
+  `AsyncResult<T, E>[]` collapses to `Result<T[], E>` / `AsyncResult<T[], E>`, and
+  a record `{ a: Result<A, E>, b: Result<B, E> }` becomes `Result<{ a: A; b: B },
+E>` (named parallel work, no tupling). All forms still short-circuit on the
+  first `Err`; `allAsync` resolves its inputs concurrently (order preserved) and
```
[the remark this answers is not in the document]
Comment (btravers) [reply CLAUDE.md]: Fixed in ccfe1dd — rewrote that CLAUDE.md bullet so the continuation no longer breaks mid-type.

File: packages/core/src/aggregate.spec.ts [hunk 6]
```diff
@@ -103,4 +121,21 @@ describe("allAsync", () => {
...
+  });
+
+  it("short-circuits a record on the first Err", async () => {
+    const r = await allAsync({
+      a: fromSafePromise(Promise.resolve(1)),
+      b: fromPromise(Promise.reject("bad"), (c) => c as string),
+    });
+    expect(r.unwrapErr()).toBe("bad");
+  });
 });
```
[the remark this answers is not in the document]
Comment (btravers) [reply packages/core/src/aggregate.spec.ts:166]: Done in ccfe1dd — added the missing parity tests for the record async path: Defect-dominates and empty-record, plus a `__proto__` pollution test (now under `allFromDictAsync`/`allFromDict`).
